tools/serverpod_cli/test/analyzer/entities/stateful_analyzer/protocol_validation/relation_database_action_test.dart

```diff
@@ -0,0 +1,388 @@
+import 'package:serverpod_cli/src/analyzer/entities/definitions.dart';
+import 'package:serverpod_cli/src/analyzer/entities/stateful_analyzer.dart';
+import 'package:serverpod_cli/src/generator/code_generation_collector.dart';
+import 'package:serverpod_cli/src/test_util/builders/protocol_source_builder.dart';
+import 'package:test/test.dart';
+
+void main() {
+  var databaseActions = [
+    'Cascade',
+    'NoAction',
+    'Restrict',
+    'SetNull',
+    'SetDefault',
+  ];
+
+  for (var action in databaseActions) {
+    group(
+        'Given a class with onUpdate database action explicitly set to $action',
+        () {
+      var protocols = [
+        ProtocolSourceBuilder().withYaml(
+          '''
+        class: Example
+        table: example
+        fields:
+          example: Example?, relation(onUpdate=$action)
+        ''',
+        ).build(),
+      ];
+
+      var collector = CodeGenerationCollector();
+      var analyzer = StatefulAnalyzer(protocols, onErrorsCollector(collector));
+      var definitions = analyzer.validateAll();
+      var entity = definitions.first as ClassDefinition;
+
+      test('then no errors are detected.', () {
+        expect(collector.errors, isEmpty);
+      });
+
+      var field = entity.findField('exampleId');
+
+      var noneFieldRelation =
+          field == null || field.relation is! ForeignRelationDefinition;
+      test('then onUpdate is set to $action.', () {
+        var relation = field?.relation as ForeignRelationDefinition;
+
+        expect(
+          relation.onUpdate.name.toString().toLowerCase(),
+          action.toLowerCase(),
+        );
+      }, skip: noneFieldRelation);
+    });
+  }
```

Comment:
I have added 3 test abstractions like this in this refactor.

Normally I would argue that abstracting stuff in tests is bad practice. But in this situation, I think it can be helpful, the point here is we don't really have any logic in the tests. The test is still declarative and each case will still execute even if one "expect" fails in one of the tests.